### PR TITLE
Fix issue #1258

### PR DESCRIPTION
**Pull request by bombless**
Fix https://github.com/rust-lang/cargo/issues/1258
Retry for up to 3 times, as @retep998 suggests

### CODE REVIEW

**Comment by bombless** [comment]
The problem itself is hard to reproduce, so I'm not sure there's good way to test the change.

**Comment by wycats** [comment]
@bombless it seems like this error should be formatted in the same way as the other status messages. Can you do that?

**Comment by bombless** [comment]
@wycats Okay, done. Is it good enough now?

File: src/cargo/sources/registry.rs [hunk 1]
```diff
@@ -524,10 +524,19 @@ impl<'a, 'b> Source for RegistrySource<'a, 'b> {
...
             url.path_mut().unwrap().push("download".to_string());
-            let path = try!(self.download_package(package, &url).chain_error(|| {
+            let mut path = self.download_package(package, &url).chain_error(|| {
                 internal(format!("Failed to download package `{}` from {}",
                                  package, url))
-            }));
+            });
+            for _ in 0 .. 3 {
+                if path.is_ok() { break }
+                try!(self.config.shell().status("Refetching", package));
```
Comment (bombless) [inline src/cargo/sources/registry.rs]: Maybe `try!(self.config.shell().err(format!("Refetching {}", package)));` is a better idea.
I'm not sure if this should be considered an error or normal status, @wycats .

Comment (bombless) [inline src/cargo/sources/registry.rs]: Or maybe ` try!(self.config.shell().warn(format!("Refetching {}", package))); `

Comment (bombless) [inline src/cargo/sources/registry.rs]: Now I think we could use `Downloading(1)`, `Downloading(2)`, etc to represent retrying.

**Comment by alexcrichton** [comment]
Hm I'm not sure that this is the best way to go about this. I would not expect that we would try *by default* to retry downloads of a crate. Additionally, this is discarding all information about the error which happened, possibly retrying for cases which will obviously never succeed. I think that https://github.com/rust-lang/cargo/issues/1258 specifically should be fixed by https://github.com/rust-lang/cargo/issues/1333 instead of retrying.

Thanks for the PR though!

**Pull request closed without merging**